xds/src/main/java/io/grpc/xds/GoogleCloudToProdNameResolver.java

```diff
@@ -0,0 +1,260 @@
+/*
+ * Copyright 2021 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Charsets;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.io.CharStreams;
+import io.grpc.NameResolver;
+import io.grpc.NameResolverRegistry;
+import io.grpc.Status;
+import io.grpc.SynchronizationContext;
+import io.grpc.alts.InternalCheckGcpEnvironment;
+import io.grpc.internal.GrpcUtil;
+import io.grpc.internal.SharedResourceHolder;
+import io.grpc.internal.SharedResourceHolder.Resource;
+import io.grpc.xds.XdsNameResolverProvider.XdsClientPoolFactory;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.net.HttpURLConnection;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.util.concurrent.Executor;
+
+/**
+ * CloudToProd version of {@link NameResolver}.
+ */
+final class GoogleCloudToProdNameResolver extends NameResolver {
+
+  @VisibleForTesting
+  static final String METADATA_URL_ZONE =
+      "http://metadata.google.internal/computeMetadata/v1/instance/zone";
+  @VisibleForTesting
+  static final String METADATA_URL_SUPPORT_IPV6 =
+      "http://metadata.google.internal/computeMetadata/v1/instance/network-interfaces/0/ipv6s";
+  @VisibleForTesting
+  static boolean isOnGcp = InternalCheckGcpEnvironment.isOnGcp();
+  @VisibleForTesting
+  static boolean xdsBootstrapProvided =
+      System.getenv("GRPC_XDS_BOOTSTRAP") != null
+          || System.getProperty("io.grpc.xds.bootstrap") != null
+          || System.getenv("GRPC_XDS_BOOTSTRAP_CONFIG") != null
+          || System.getProperty("io.grpc.xds.bootstrapValue") != null;
+
+  private HttpConnectionProvider httpConnectionProvider = HttpConnectionFactory.INSTANCE;
+  private final String authority;
+  private final SynchronizationContext syncContext;
+  private final Resource<Executor> executorResource;
+  private final XdsClientPoolFactory xdsClientPoolFactory;
+  private final NameResolver delegate;
+  private final boolean usingExecutorResource;
+  private final String schemeOverride = !isOnGcp || xdsBootstrapProvided ? "dns" : "xds";
```

Comment:
It indeed is. That's also something I asked Mark. It's weird, but this is indeed the logic: https://github.com/grpc/grpc/blob/3944bfaf76b3123d40c68d29e3c1eeeed08afe57/src/core/ext/filters/client_channel/resolver/google_c2p/google_c2p_resolver.cc#L234-L240

Reply:
OK, I do see
```
Therefore, if the GRPC_XDS_BOOTSTRAP and GRPC_XDS_BOOTSTRAP_CONFIG env vars are defined, the C2P resolver will delegate to the DNS resolver (i.e., it will basically replace the "google-c2p" URI scheme with "dns" and create a new resolver for the new URI).  This will effectively eliminate the use of DirectPath C2P for PSM users; all of their C2P traffic will be sent via the CFEs.
```

in the spec. It's confusing to read at first glance.

Reply:
It is indeed very confusing. The first time I read it, I thought there was a typo...

Reply:
Worth a comment here, so people won't be confused again.